Replace compiler with language flag if it's a C++ compiler.

clin18 opened the pull request:
Currently, if we have a header file for which its corresponding C++ file exists
in the compilation database, we forget that it's supposed to be a C++ header by
stripping away the compiler.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/248)
<!-- Reviewable:end -->

puremourning commented:
I think we need to add some module tests for this. 
<!-- Reviewable comment -K-mOJycvZh51yJMbeOj:-668651856 -->

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

Valloric commented:
Thanks for sending a PR! :)

Agreed with @puremourning, tests needed. Also minor other comments added.
<!-- Reviewable comment -K-zpj63j4BSQp_cvSkQ:-1716117413 -->

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[ycmd/completers/cpp/flags.py, line 197 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpP0Oysv3VlMtu07M)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L197)):</sup>
This should probably be named `first_flag`, not `compiler`. It might not contain a compiler name.
<!-- Reviewable comment -K-zpP0Oysv3VlMtu07N:1914194888 -->

---

<sup>**[ycmd/completers/cpp/flags.py, line 200 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpa1IkPq74ZRXVvbr)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L200)):</sup>
This needs a comment explaining why `++` is magic (I understand why, but the next reader might not).
<!-- Reviewable comment -K-zpa1JFqBMgZH4VENg:-1742656018 -->

---

<sup>**[ycmd/completers/cpp/flags.py, line 201 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpgQqOzCQAx4XLH2N)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L201)):</sup>
You need tests for this behavior.
<!-- Reviewable comment -K-zpgQqOzCQAx4XLH2O:-328577740 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

clin18 commented:
Addressed comments and added tests with 954f782. Please take another look.

Some questions:
- There's a clang_helpers.py that contains similar code, should that be updated as well?
- Does it make sense to add a C language flag in case it's *not* a C++ compiler? E.g., `clang foo.cpp` should force `foo.cpp` to be parsed as a C file. I'm not sure if this would interfere with how YCM handles other languages (e.g., Objective-C, C#, etc)

clin18 commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[ycmd/completers/cpp/flags.py, line 197 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpP0Oysv3VlMtu07M)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L197)):</sup>
Done.
<!-- Reviewable comment -K0-NA-TTXVsb2NrA3rQ:-499146009 -->

---

<sup>**[ycmd/completers/cpp/flags.py, line 200 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpa1IkPq74ZRXVvbr)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L200)):</sup>
Done.
<!-- Reviewable comment -K0-N4S-obzepZJJZhMd:-499146009 -->

---

<sup>**[ycmd/completers/cpp/flags.py, line 201 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K-zpgQqOzCQAx4XLH2N)** ([raw file](https://github.com/valloric/ycmd/blob/95d0b356c901f516dee5bfa91eae5d44c8ff85c9/ycmd/completers/cpp/flags.py#L201)):</sup>
Done.
<!-- Reviewable comment -K0-N56YpS360SYX9FnO:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

Valloric commented:
Assuming `clang foo.cpp` means C code can be tricky. What if the user also passes `-x c++` as well to such an invocation? I'm assuming clang will do the right thing. If not, then it might make sense to do this.

WRT `clang_helpers.py`, that file is loaded directly by old `ycm_extra_conf.py` files. It's been obsolete for a while, but we keep it around for the sake of backwards compatibility. Probably best not to touch it.
<!-- Reviewable comment -K09DpKaaJtShMdiShNe:-320478433 -->

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

clin18 commented:
`clang -x c++ foo.cpp` would turn into `-x c -x c++ foo.cpp`. This is why I replaced the flag at the beginning.

Valloric commented:
Maybe we should add `-x c` then. Sounds like it wouldn't break users.
<!-- Reviewable comment -K09KBunu8mUGwfZ5nty:1467297567 -->

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[ycmd/completers/cpp/flags.py, line 195 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K0D3s0tm4OyK7rUG_5M)** ([raw file](https://github.com/valloric/ycmd/blob/e80428fa51d19d721e44c70c9893611d2b3d8980/ycmd/completers/cpp/flags.py#L195)):</sup>
Maybe the docstring should talk about the use of the C language flag if not a C++ compiler.
<!-- Reviewable comment -K0D2Cw5pcEVkiBCnBly:645683519 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

clin18 commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[ycmd/completers/cpp/flags.py, line 195 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/248#-K0D3s0tm4OyK7rUG_5M)** ([raw file](https://github.com/valloric/ycmd/blob/e80428fa51d19d721e44c70c9893611d2b3d8980/ycmd/completers/cpp/flags.py#L195)):</sup>
Done.
<!-- Reviewable comment -K0eC1z2oVQVzXwDy5gM:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

Valloric commented:
This seems ready to go; thanks for the PR!

@homu r+
<!-- Reviewable comment -K0wKo2BVxOzoZqMiomg:-1413788931 -->

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/248)
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 9fb379e has been approved by `Valloric`

<!-- @homu r=Valloric 9fb379e37ece1eb9bf8021e3af719c6a1af5eef3 -->

homu commented:
:hourglass: Testing commit 9fb379e with merge 0825f85...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/86066420)
